src/main/java/com/conveyal/datatools/manager/controllers/api/FeedSourceController.java

```diff
@@ -40,130 +41,118 @@
+            for (String resourceType : DataManager.feedResources.keySet()) {
+                DataManager.feedResources.get(resourceType).feedSourceCreated(newFeedSource, req.headers("Authorization"));
+            }
+            // Notify project subscribers of new feed source creation.
+            Project parentProject = Persistence.projects.getById(newFeedSource.projectId);
+            NotifyUsersForSubscriptionJob.createNotification(
+                "project-updated",
+                newFeedSource.projectId,
+                String.format("New feed %s created in project %s.", newFeedSource.name, parentProject.name));
+            return newFeedSource;
+        } catch (Exception e) {
+            logMessageAndHalt(req, 500, "Unknown error encountered creating feed source", e);
+            return null;
+        }
     }
 
-    public static FeedSource updateFeedSource(Request req, Response res) {
+    /**
+     * Spark HTTP endpoint to update a feed source. Note: at one point this endpoint accepted a JSON object
+     * representing a single field to update for the feed source, but it now requires that the JSON body represent all
+     * fields the updated feed source should contain. This change allows us to parse the JSON into a POJO, which
+     * essentially does type checking for us and prevents issues with deserialization from MongoDB into POJOs.
+     */
+    private static FeedSource updateFeedSource(Request req, Response res) throws IOException {
         String feedSourceId = req.params("id");
-
-        // call this method just for null and permissions check
-        // TODO: it's wasteful to request the entire feed source here, need to factor out permissions checks. However,
-        // we need the URL to see if it has been updated in order to then set the lastFetched value to null.
         FeedSource formerFeedSource = requestFeedSourceById(req, Actions.MANAGE);
-        Document fieldsToUpdate = Document.parse(req.body());
-        if (fieldsToUpdate.containsKey("url") && formerFeedSource.url != null) {
-            // Reset last fetched timestamp if the URL has been updated.
-            if (!fieldsToUpdate.get("url").toString().equals(formerFeedSource.url.toString())) {
-                LOG.info("Feed source fetch URL has been modified. Resetting lastFetched value from {} to {}", formerFeedSource.lastFetched, null);
-                fieldsToUpdate.put("lastFetched", null);
-            }
+        FeedSource updatedFeedSource = getPOJOFromRequestBody(req, FeedSource.class);
+        // Feed source previously had a URL, but it has been changed. In this case, we reset the last fetched timestamp.
+        if (formerFeedSource.url != null && !formerFeedSource.url.equals(updatedFeedSource.url)) {
+            LOG.info("Feed source fetch URL has been modified. Resetting lastFetched value from {} to {}", formerFeedSource.lastFetched, null);
+            updatedFeedSource.lastFetched = null;
```

Comment:
Are we sure we need to do this? Seems like we'd lose valuable information about when a feed was last fetched even if it was from an old url.

Reply:
Do you have an alternative approach in mind, @evansiroky?

Reply:
Seems like it'd make sense to store and retrieve that information from the most recent Feed Version. That might require adding additional information to the Feed Version model though.

Reply:
This is not really new code, so if you think it's worth pulling a ticket for to track and continue the conversation, feel free. I'm not sure we should make these changes in this context though.